enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,239 @@
+
+### Implementation Details/Notes/Constraints
+
+#### Image building
+
+ICC will follow different code paths depending on whether the `BareMetalHost`
+has the `infraenvs.agent-install.openshift.io` annotation linking it to an
+`InfraEnv` resource. In both cases it will generate an Ignition configuration
+to start and configure the Ironic agent. In the ZTP case, the corresponding
+`InfraEnv` will be updated with the generated Ignition configuration (merging
+it with any user-provided configuration).
+
+The same process will be applied to `initrd` images for (i)PXE booting. The
+Ignition configuration will be embedded into the resulting `initrd` by the
+assisted service. The root filesystem (required for CoreOS) will be provided
+by [coreos-downloader][coreos-downloader], similarly to the *standard* flow.
+
+#### Coordination between agents
+
+The first agent to start will be the Ironic agent since it is responsible
+for coordinating the whole flow and (indirectly through Ironic) updating the
+`BareMetalHost`'s status. *Inspection* and additional configuration (RAID etc)
+will be handled by the Ironic agent alone.
+
+Currently, the installation is initiated by applying the following change to
+the `BareMetalHost`'s spec:
+
+```yaml
+Image:
+  URL: url/of/assisted/agent.iso
+  DiskFormat: live-iso
+```
+
+The new process will be instead triggered by:
+
+```yaml
+CustomDeploy:
+  Method: install_assisted
+```
+
+This instructs Ironic to find and execute a *custom deploy step* called
+`install_assisted`, which will be provided by our Ironic agent plugin. This
+step will use D-Bus to order systemd to start the assisted agent and watch
+its status.
+
+If the assisted agent exits successfully, the Ironic agent will report success
+to Ironic and baremetal-operator. Otherwise, a failure will be recorded.
+
+### Risks and Mitigations
+
+- More containers will be downloaded on the ramdisk start-up. We'll need to
+  account for more RAM and a longer start-up time.
+- A user with editing access to `InfraEnv` resources can potentially break
+  the Ironic agent configuration. Short-term, we'll probably document this
+  as a known issue. Long-term, we may need to update `InfraEnv` with a new
+  field.
+
+## Design Details
+
+### Open Questions
+
+- We are going to put the user's pull secret directly to the `InfraEnv`
+  resources as part of the Ignition override. Is it acceptable?
+- Is there a chance of version mismatch between the CoreOS `initrd` from the
+  assisted service and the root filesystem provided by
+  [coreos-downloader][coreos-downloader]?
+
+### Test Plan
+
+The feature will be covered by the existing ZTP e2e tests.
+
+### Graduation Criteria
+
+N/A
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+This change does not introduce new resources, so upgrading or downgrading them
+is not a concern.
```

Comment:
The linkage is via labels, which are already set (by BMAC or a user - Flavio may know that).

Reply:
I assume we're excluding the upgrade impacts on the BMAC from this enhancement?

Reply:
At the very least, someone else will have to write it. I don't understand BMAC even remotely well enough.

Reply:
Outlining the BMAC upgrade impact (and how we'll ensure test coverage) seems like a good idea, e.g we'll need to be very sure that BMAC leaves existing deployed BMH resources alone, that scale-out post upgrade works as expected, etc - @flaper87 can you help with that?

Reply:
_Something_ will still have to set the CustomDeploy method at provisioning time. I don't recall where we got to on which part of assisted-service is expected to do that.

Reply:
> Something will still have to set the CustomDeploy method at provisioning time. I don't recall where we got to on which part of assisted-service is expected to do that.

Fair point, I knew I was forgetting something. Who sets it in the `install_coreos` case? The user? If so, then we can let the user do this for assisted to. Otherwise, BMAC or ICC can do it.

Reply:
> Who sets it in the `install_coreos` case?

CAPBM